pkg/istio/authorizationpolicy/controller.go

```diff
@@ -0,0 +1,549 @@
+			if err != nil {
+				continue
+			}
+			// if svc match with current svc, process it and add it to the rules
+			// note that svc defined on athenz can be a regex, need to match the pattern
+			res, e := regexp.MatchString(svc, svcLabel)
+			if e != nil {
+				log.Errorln("error matching string: ", e.Error())
+			}
+			if res {
+				rule := &v1beta1.Rule{}
+
+				// form rule.From, must initialize internal source here
+				from_principal := &v1beta1.Rule_From{
+					Source: &v1beta1.Source{},
+				}
+				from_requestPrincipal := &v1beta1.Rule_From{
+					Source: &v1beta1.Source{},
+				}
+				// role name should match zms resource name
+				for _, roleName := range athenzModel.Roles {
+					if roleName == role {
+						// add function to enableOriginJwtSubject, following code assume enableOriginJwtSubject is true by default
+						if c.enableOriginJwtSubject {
+							for _, roleMember := range athenzModel.Members[roleName] {
+								spiffeName, err := memberToSpiffe(roleMember)
+								if err != nil {
+									log.Errorln("error converting role name to spiffeName: ", err.Error())
+									continue
+								}
+								from_principal.Source.Principals = append(from_principal.Source.Principals, spiffeName)
+								originJwtName, err := memberToOriginJwtSubject(roleMember)
+								if err != nil {
+									log.Errorln(err.Error())
+									continue
+								}
+								from_requestPrincipal.Source.RequestPrincipals = append(from_requestPrincipal.Source.RequestPrincipals, originJwtName)
+							}
+						}
+						//add role spiffee for role certificate
+						roleSpiffeName, err := RoleToSpiffe(string(athenzModel.Name), string(roleName))
+						if err != nil {
+							log.Println("error when convert role to spiffe name: ", err.Error())
+							continue
+						}
+						from_principal.Source.Principals = append(from_principal.Source.Principals, roleSpiffeName)
+					}
+				}
+				rule.From = append(rule.From, from_principal)
+				rule.From = append(rule.From, from_requestPrincipal)
+				// form rules_to
+				rule_to := &v1beta1.Rule{}
+				_, err = parseAssertionEffect(assert)
+				if err != nil {
+					log.Debugf(err.Error())
+					continue
+				}
+				method, err := parseAssertionAction(assert)
+				if err != nil {
+					log.Debugf(err.Error())
+					continue
+				}
+				// form rule.To
+				to := &v1beta1.Rule_To{
+					Operation: &v1beta1.Operation{
+						Methods: []string{method},
+					},
+				}
+				if path != "" {
+					to.Operation.Paths = []string{path}
+				}
+				rule_to.To = append(rule_to.To, to)
+				rules = append(rules, rule_to)
+				rules = append(rules, rule)
+			}
+		}
+	}
+	spec.Rules = rules
+	out.Spec = spec
+	return out
+}
+
+// parseAssertionResource parses the resource of an action into the service name (AccessRule constraint) and the
+// HTTP paths if specified (suffix :<path>)
+func parseAssertionResource(domainName zms.DomainName, assertion *zms.Assertion) (string, string, error) {
```

Comment:
Can we reuse all the functions in the common package instead of having them here? We can move them to different files if they're currently in the service role / binding files.

Reply:
moved to a common package